evaluation.py

```diff
@@ -73,39 +95,56 @@ def eval(model, loss_function, test_iter, args):
     model.train(mode)
     return eval_results
 
-def greedy_decoding(model, test_iter, args):
+def _idxs_to_sentences(idxss):
+    raw_sentences = [[model.trg_vocab.itos[i] for i in idxs] for idxs in idxss]
+    sentences = []
+    for rs in raw_sentences:
+        s = []
+        for tok in rs:
+            if tok in (BOS_TOKEN, PAD_TOKEN):
+                continue
+            elif tok == EOS_TOKEN:
+                break
+            else:
+                s.append(tok)
+        sentences.append(" ".join(s))
+    return sentences
+
+
+def greedy_decoding(model, test_iter, max_decoding_len):
     mode = model.training
     model.eval()
 
     ground_truth = []
     predictions = []
-    EOS_IDX = model.trg_vocab.stoi(EOS_TOKEN)
-    PAD_IDX = model.trg_vocab.stoi(PAD_TOKEN)
-    vocab_id = torch.eye(len(model.trg_vocab))
     with torch.no_grad():
         for batch in tqdm(test_iter):
-            still_decoding = True
-            current_time_step = 0
-            decoded_sentences = torch.zeros(batch.trg.shape) # t x b x v
-            decoded_sentences[:,:,PAD_IDX] = 1 # Transform the entire sentence into pads
-            ground_truth += batch.trg.tolist()
-            while current_time_step < args['max_decoding_len'] and still_decoding:
-                logits = model(batch.src, decoded_sentences)
-                decoded_sentences[current_time_step,:,:] = vocab_id[logits.argmax(-1)]
-                current_time_step += 1
-                still_decoding = torch.all(torch.any(torch.argmax(decoded_sentences, dim = -1) == EOS_IDX, dim = 0))
-            predictions += list(logits.argmax(-1).transpose(1,0).tolist())
+            ground_truth += batch.trg.transpose(1, 0).tolist()
+            predictions += model.decode(batch.src, max_decoding_len)
+
+    # TODO What is this?
+    # model.train(mode)
```

Comment:
Can you put this back?
It will revert the eval/train flag of the model to the original mode (so if the model was in training mode it will revert to that). I think it helps to know that the functions will not alter the state of the model.